vms/platformvm/state/merkle_state.go

```diff
@@ -0,0 +1,1530 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package state
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/ava-labs/avalanchego/cache"
+	"github.com/ava-labs/avalanchego/cache/metercacher"
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/database/prefixdb"
+	"github.com/ava-labs/avalanchego/database/versiondb"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/trace"
+	"github.com/ava-labs/avalanchego/utils"
+	"github.com/ava-labs/avalanchego/utils/constants"
+	"github.com/ava-labs/avalanchego/utils/crypto/bls"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/utils/wrappers"
+	"github.com/ava-labs/avalanchego/vms/components/avax"
+	"github.com/ava-labs/avalanchego/vms/platformvm/blocks"
+	"github.com/ava-labs/avalanchego/vms/platformvm/config"
+	"github.com/ava-labs/avalanchego/vms/platformvm/metrics"
+	"github.com/ava-labs/avalanchego/vms/platformvm/reward"
+	"github.com/ava-labs/avalanchego/vms/platformvm/status"
+	"github.com/ava-labs/avalanchego/vms/platformvm/txs"
+	"github.com/ava-labs/avalanchego/x/merkledb"
+)
+
+const (
+	HistoryLength = int(256)    // from HyperSDK
+	NodeCacheSize = int(65_536) // from HyperSDK
+
+	utxoCacheSize = 8192 // from avax/utxo_state.go
+)
+
+var (
+	_ State = (*merkleState)(nil)
+
+	merkleStatePrefix      = []byte{0x0}
+	merkleSingletonPrefix  = []byte{0x1}
+	merkleBlockPrefix      = []byte{0x2}
+	merkleTxPrefix         = []byte{0x3}
+	merkleIndexUTXOsPrefix = []byte{0x4} // to serve UTXOIDs(addr)
+	merkleUptimesPrefix    = []byte{0x5} // locally measured uptimes
+	merkleWeightDiffPrefix = []byte{0x6} // non-merklelized validators weight diff. TODO: should we merklelize them?
```

Comment:
if not, a state-synced node which has downloaded the merkleDB will have to wait some time to get ready? Can't we rebuild diffs by downloading enough blocks, following mekle state download, similarly to what we do with C-chain?